test/jdk/jdk/internal/ref/Cleaner/CleanableListTest.java

```diff
@@ -0,0 +1,129 @@
+ * @modules java.base/jdk.internal.ref
+ * @run testng/othervm CleanableListTest
+ */
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Random;
+
+import jdk.internal.ref.CleanableListTestHelper;
+import jdk.internal.ref.TestCleanable;
+import jdk.test.lib.RandomFactory;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.testng.annotations.Ignore;
+
+public class CleanableListTest {
+
+    static final int SINGLE_NODE_CAPACITY = CleanableListTestHelper.NODE_CAPACITY;
+    static final int MULTI_NODE_CAPACITY = CleanableListTestHelper.NODE_CAPACITY * 4;
+
+    static final Random RND = RandomFactory.getRandom();
+    static final int RANDOM_ITERATIONS = 10_000_000;
+
+    @Test
+    public void testSingle() {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+        TestCleanable tc = list.newCleanable();
+        Assert.assertFalse(list.isEmpty());
+        Assert.assertTrue(list.remove(tc));
+        Assert.assertTrue(list.isEmpty());
+        Assert.assertFalse(list.remove(tc));
+    }
+
+    @Test
+    public void testSequential_Single() {
+        doSequential(SINGLE_NODE_CAPACITY);
+    }
+
+    @Test
+    public void testSequential_Multi() {
+        doSequential(MULTI_NODE_CAPACITY);
+    }
+
+    private void doSequential(int size) {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+
+        List<TestCleanable> tcs = new ArrayList<>();
+        for (int c = 0; c < size; c++) {
+            tcs.add(list.newCleanable());
+        }
+        Assert.assertFalse(list.isEmpty());
+
+        for (TestCleanable tc : tcs) {
+            Assert.assertTrue(list.remove(tc));
+        }
+        Assert.assertTrue(list.isEmpty());
+    }
+
+    @Test
+    public void testRandom_Single() {
+        doRandom(SINGLE_NODE_CAPACITY);
+    }
+
+    @Test
+    public void testRandom_Multi() {
+        doRandom(MULTI_NODE_CAPACITY);
+    }
+
+    private void doRandom(int size) {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+
+        BitSet bs = new BitSet(size);
+
+        List<TestCleanable> tcs = new ArrayList<>();
+        for (int c = 0; c < size; c++) {
+            tcs.add(list.newCleanable());
+            bs.set(c, true);
+        }
+        Assert.assertFalse(list.isEmpty());
+
+        for (int t = 0; t < RANDOM_ITERATIONS; t++) {
+            int idx = RND.nextInt(size);
+            TestCleanable tc = tcs.get(idx);
+            if (bs.get(idx)) {
+                Assert.assertTrue(list.remove(tc));
+                bs.set(idx, false);
+            } else {
+                Assert.assertFalse(list.remove(tc));
+                list.insert(tc);
+                bs.set(idx, true);
+            }
+        }
```

Comment:
Can the test remove everything left in the list after doing the `RANDOM_ITERATIONS`?

Reply:
Done in new commit.